### DISCUSSION TITLE
Would it be possible to bind the junit jupiter ExtensionContext into the applicaton injector?

### DISCUSSION THREAD

**Discussion by JKomoroski** [xvik/dropwizard-guicey#388] [General]
Sometimes when I'm writing a configuration hook, I like to bind overrrides from a test module. It would be nice if I could obtain data out of the test extension context when writing those provider methods:

```
MyTestModule extends AbstractModule implements GuiceyConfigurationHook {
    
    @Override
    public void configure(Builder builder) {
        builder.modulesOverride(this);
    }
    
    @Provides
    @Singleton
    MyObject provideOverrideSqsClient(ExtensionContext extensionContext) { // bound when running tests with jupiter
        return MyObjectExtension.get(extensionContext);
    }
 }
 ```

**Comment by xvik** [comment]
Thank you for the idea! I will look this later.

**Reply by JKomoroski** [reply]
I am not certain exactly how this would work with all the different jupiter test lifecycles. it might be that Singleton Scope and per method lifecycle doesn't work right. I'm not certain exactly which step the injector is initialized.

**Comment by xvik** [comment]
Sorry for delay. In the upcoming version, junit context could be obtained within `TestEnvioronmentSetup`. Note that guicey hooks used not only for tests, but could be used in application runtime and so they can't know anything about tests. `TestEnvioronmentSetup` is like a hook, but with test knowledge. Plus, setup object could register it's own hooks (executed before hooks).

For example (field declaration):

```java
public class MyTest {

    @EnableSetup
    TestEnvironmentSetup setup  =  (extension) -> {
        // entire junit context (class or method, it depends)
        final ExtensionContext context = extension.getJunitContext();

         // next you can register hook with required data from context:
         extension.hooks((builder)->{
               builder.modulesOverride(new MyTestModule(context))
         });
    }
}
```

Another variant (class declaration):

```java
public class MySetup extends AbstractModule implements TestEnvironmentSetup, GuiceyConfigurationHook {
 
     private ExtensionContext context;
   
    @Override
    public Object setup(TestExtension extension) {
          context = extension.getJunitContext();
          // register itself as a hook
          extension.hooks(this);
          return null;
    }

    @Override
    public void configure(GuiceBundle.Builder builder) {
           // register itself as module
           builder.modulesOverride(this);
    }
   
    // optional
    @Override
    public void configure(Builder builder) {
        builder.modulesOverride(this);
    }

    @Provides
    @Singleton
    MyObject provideOverrideSqsClient() {
        return MyObjectExtension.get(context);
    }
}
```

```java
@TestGuiceyApp(value = App.class, setup = MySetup.class)
```

I suppose this should be enough for your needs. If you want you can easilly bind entire junit context into guice context.

**Comment by xvik** [comment]
[7.2.0 released](https://xvik.github.io/dropwizard-guicey/7.2.0/about/release-notes/). As mentioned before, `TestEnvironmentSetup` now could access junit `ExtensionContext`:

```java
    @EnableSetup
    TestEnvironmentSetup setup  =  (extension) -> {
        // entire junit context (class or method, it depends)
        final ExtensionContext context = extension.getJunitContext();

       extension.hooks((builder)->{
               builder.modulesOverride(new MyTestModule(context))
       });
}
```

Also, context could be injected as a test method parameter:

```java
@BeforeEach
public void setUp(ExtensionContext context) {
}
```

Also, [setup objects were highly improved](https://xvik.github.io/dropwizard-guicey/7.2.0/guide/test/junit5/setup-object/) and now could replace custom junit extensions completely.
